realm/realm-library/src/androidTestObjectServer/java/io/realm/SyncedRealmMigrationTests.java

```diff
@@ -56,4 +64,186 @@ public void migrateRealm_syncConfigurationThrows() {
+        dynamicRealm.close();
+
+        try {
+            Realm.getInstance(config);
+            fail();
+        } catch (IllegalStateException ignored) {
+        }
+    }
+
+    // Check that indexes are not being added if the schema version is the same
+    @Test
+    public void sameSchemaVersion_doNotRebuildIndexes() {
+
+        SyncConfiguration config = configFactory.createSyncConfigurationBuilder(SyncTestUtils.createTestUser(), "http://foo.com/auth")
+                .schema(IndexedFields.class)
+                .schemaVersion(42)
+                .build();
+
+        // Setup initial Realm schema (with no indexes)
+        String className = IndexedFields.class.getSimpleName();
+        DynamicRealm dynamicRealm = DynamicRealm.getInstance(config);
+        RealmSchema schema = dynamicRealm.getSchema();
+        dynamicRealm.beginTransaction();
+        schema.create(className)
+                .addField(IndexedFields.FIELD_INDEXED_STRING, String.class) // No index
+                .addField(IndexedFields.FIELD_NON_INDEXED_STRING, String.class);
+        dynamicRealm.setVersion(42);
+        dynamicRealm.commitTransaction();
+        dynamicRealm.close();
+
+        try {
+            Realm realm = Realm.getInstance(config); // Opening at same schema version (42) will not rebuild indexes
+            fail();
+        } catch (RealmMigrationNeededException ignored) {
+        }
+
+// FIXME: This is the intended behaviour
+//        RealmObjectSchema indexedFieldsSchema = realm.getSchema().get(className);
+//        try {
+//            assertFalse(indexedFieldsSchema.hasIndex(IndexedFields.FIELD_INDEXED_STRING));
+//            assertFalse(indexedFieldsSchema.hasIndex(IndexedFields.FIELD_NON_INDEXED_STRING));
+//        } finally {
+//            realm.close();
+//        }
+    }
+
+    // Check that indexes are being added if the schema version is different
+    @Test
+    public void differentSchemaVersions_rebuildIndexes() {
+
+        SyncConfiguration config = configFactory.createSyncConfigurationBuilder(SyncTestUtils.createTestUser(), "http://foo.com/auth")
+                .schema(IndexedFields.class)
+                .schemaVersion(42)
+                .build();
+
+        // Setup initial Realm schema (with no indexes)
+        String className = IndexedFields.class.getSimpleName();
+        DynamicRealm dynamicRealm = DynamicRealm.getInstance(config);
+        RealmSchema schema = dynamicRealm.getSchema();
+        dynamicRealm.beginTransaction();
+        schema.create(className)
+                .addField(IndexedFields.FIELD_INDEXED_STRING, String.class) // No index
+                .addField(IndexedFields.FIELD_NON_INDEXED_STRING, String.class);
+        dynamicRealm.setVersion(43);
+        dynamicRealm.commitTransaction();
+        dynamicRealm.close();
+
+        try {
+            Realm realm = Realm.getInstance(config); // Opening at different schema version (42) should rebuild indexes
+            fail();
```

Comment:
Nothing, they will be in deep trouble, because this might happen at runtime. It is quite a big bug we should fix as soon as possible, but it is outside the scope of the PR (and luckily no-one hit it yet).

To work around it right now you need to define all indexes up front just like primary keys and they cannot be changed later.